### PR TITLE
Baseline Experiments

### PR DESCRIPTION
**Pull request by rpanah**
This is the baseline URL testing experiment. It does HTTP GET, DNS, TLS cert get, and traceroute on mutiple lists of URLs while recording pcaps for each URL. It has been tested a few times and enhanced in order to take less time to finish and consume less disk space.
@ben-jones, can you please review this?

### CODE REVIEW

File: centinel/client.py [hunk 1]
```diff
@@ -151,9 +167,29 @@ def run(self, data_dir=None):
...
+                # they could be anything that doesn't belong in the json file
+                # (e.g. pcap files)
+                # these should all be compress with bzip2
+                # the experiment is responsible for giving these a name and
+                # keeping a list of files in the json results
+                if exp.external_results is not None:
+                    for fname, fcontents in exp.external_results.items():
+                        external_file_name = "external_%s-%s-%s.bz2" % (name,
+                            exp_start_time,fname)
+                        external_file_path = os.path.join(
```
Comment (ben-jones) [inline centinel/client.py]: more formatting issues

File: centinel/client.py [hunk 2]
```diff
@@ -163,7 +199,7 @@ def run(self, data_dir=None):
                     logging.info("tcpdump stopped.")
                     try:
                         pcap_file_name = "pcap_%s-%s.pcap.bz2" % (name, 
-                            datetime.now().isoformat())
+                            exp_start_time)
```
Comment (ben-jones) [inline centinel/client.py]: need to indent here

Comment (rpanah) [inline centinel/client.py]: Adding more indent to match where "name," starts will make the line longer than 78 chars and violates PEP8.

File: centinel/client.py (hunk 1 above)
Comment (rpanah) [inline centinel/client.py]: Adding more indent breaks PEP8. Read comment below.

File: centinel/data/country.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+"yahoo.fr/this,and,that",xx,yy
```
Comment (ben-jones) [inline centinel/data/country.txt]: I don't think commas are legal in URLs-> you use & to distinguish between url parameters, which are key value pairs. What is the purpose of this line?

Comment (rpanah) [inline centinel/data/country.txt]: Comma's are allowed in URL paths, (they are reserved characters in domain names).
Citizen Lab's URLs are sometimes enclosed in quotes because they contain comma's in their paths.

File: centinel/experiments/baseline.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        comments = ""
+
+        # we may want to make this threaded and concurrent
+        for line in file_contents:
+            line = line.strip()
+
+            # parse file comments, if it looks like "key : value",
+            # parse it as a key-value pair. otherwise, just
+            # store it as a raw comment.
+            if line[0] == '#':
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: this *only* parses file comments. What is the goal here? You could also write much cleaner code-> e.g. tell people that if the line starts with "#", then it is a comment

File: centinel/experiments/baseline.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                else:
+                    file_comments.append(line)
+                continue
+
+            url = line
+            meta = ''
+            url_index = url_index + 1
+
+            # handle cases where URL is enclosed in quotes
+            if line[0] == '"':
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: why are we even considering this as a case? Just use a csv parser from the csv library and it will automatically interpret quotes for you-> everything inside quotes is a single item, even if there are commas in the quote

File: centinel/experiments/baseline.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+                url = line[1:next_quote + 1]
+                # try to separate metadata, if any
+                try:
+                    meta = ','.join(line[next_quote + 2:].split(',')[1:])
+                except:
+                    pass
+
+            # if the line doesn't start with a quote
+            # but the list entry has comma separated meta-data
+            elif len(line.split(',')) > 1:
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: you keep doing all of these checks. Your code would be much clearer if you made a few booleans to store whether or not it is a comment and read in the line once. You could use csv to read the line in, then just have a few sections where you process-> it would make your code more concise and cleaner

File: centinel/experiments/baseline.py (hunk 4 above)
Comment (rpanah) [inline centinel/experiments/baseline.py]: The goal here is to support metadata that is going to be included in Citizen Lab's URL lists. This info can be stored in the form of key-value pairs which will be parsed and added to the result dictionary.

>  You could also write much cleaner code

This is too general. Would you care to elaborate?

> -> e.g. tell people that if the line starts with "#", then it is a comment

I think that part is obvious. What is not so obvious, is the key-value pair comments, which I have explained.

File: centinel/client.py (hunk 2 above)
Comment (ben-jones) [inline centinel/client.py]: this is the perfect use case for a new variable or you could force the start of the arguments to the next line with something like:
("start of line goes here"
 "" % (args))

File: centinel/experiments/baseline.py (hunk 4 above)
Comment (ben-jones) [inline centinel/experiments/baseline.py]: just make the check for a comment into a boolean variable and use that in later if statements. It's clearer with low overhead

File: centinel/experiments/baseline.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                url = url.strip()
+
+            http_ssl = False
+            ssl_port = 443
+            http_path = '/'
+
+            # parse the URL to extract netlocation, HTTP path, domain name,
+            # and HTTP method (SSL or plain)
+            try:
+                http_netloc = ''.join(urlparse.urlparse(url).netloc)
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: you call urlparse multiple times on the same URL. You should do the parse once, store it an a variable, and reference it later.

File: centinel/experiments/baseline.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+                        ssl_port = http_netloc.split(':')[1]
+
+            except Exception as exp:
+                logging.warning("%s: failed to parse URL: %s" %(url, str(exp)))
+                http_netloc = url
+                http_ssl    = False
+                ssl_port = 443
+                http_path   = '/'
+
+            domain_name = http_netloc.split(':')[0]
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: is this going to work if you get an exception? Wouldn't that just give you the scheme, e.g. http, if you hit an exception? Also, you should just put this where you put the other URL parsing stuff or in the exception handling stuff

File: centinel/experiments/baseline.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            # Traceroute
+            for method in self.traceroute_methods:
+                try:
+                    logging.info("%s: Traceroute (%s)"
+                                 % (domain_name, method.upper()))
+                    traceroute_results[domain_name] = traceroute.traceroute(
+                        domain_name, method=method)
+                except Exception as exp:
+                    logging.info("%s: Traceroute (%s) failed: %s" %
+                                    (domain_name, method.upper(), str(exp)))
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: you should do consistent indentation here-> move the argument tuple back 3 or so characters

File: centinel/primitives/dnslib.py [hunk 10]
```diff
@@ -15,7 +15,7 @@ def get_ips(host, nameserver=None, record="A"):
     return lookup_domain(host, nameservers=nameservers, rtype=record)
 
 
-def lookup_domain(domain, nameservers=[], rtype="A", timeout=10):
+def lookup_domain(domain, nameservers=[], rtype="A", timeout=2):
```
Comment (ben-jones) [inline centinel/primitives/dnslib.py]: will this be sufficient in locales with poor network connectivity?

Comment (rpanah) [inline centinel/primitives/dnslib.py]: I think it is safe to say a 2000 millisecond delay in receiving DNS response is an anomaly in every network.

Comment (ben-jones) [inline centinel/primitives/dnslib.py]: I think that 2 seconds should be sufficient, but is not guaranteed to be. Some regions (think Africa) typically have RTTs in the range of 400-500ms. If this is a query over TCP (common for EDNS options), then you have two RTTs for the query, which could easily be over 1 second. It would be odd, but I could totally see congestion issues leading to longer timeouts. In our case, the timeout is defined for the socket, but I think it is possible for the timeout to exceed 2 seconds.

Ultimately, the average or typical case really doesn't matter in this case. If we are going to make that claim that a DNS request is being dropped based on this timeout, we cannot settle for an inconclusive measurement-> 2 second is not conclusive enough for me. In these odd cases, we *need* to wait longer to differentiate packet drops and longer delays. If performance is a concern, then I also built a parallel version of this, so it really doesn't matter.

tl;dr We are making claims about censorship based on these queries, so we cannot afford uncertainty. We must have a longer timeout

Comment (ben-jones) [inline centinel/primitives/dnslib.py]: Also, some of the values you provide (rtype A and the original timeout) are the default values and you don't need to set them (unless you intended that to clarify what you are doing.)

Comment (ben-jones) [inline centinel/primitives/dnslib.py]: Also, that is the default timeout I made for the dns library. If we want to discuss changing the timeout, it should be there.

Comment (rpanah) [inline centinel/primitives/dnslib.py]: > If we want to discuss changing the timeout, it should be there.

This is the library.

Comment (ben-jones) [inline centinel/primitives/dnslib.py]: Lol. yup. So, thoughts?

Comment (rpanah) [inline centinel/primitives/dnslib.py]: > Also, some of the values you provide [...]

Again, this **is** the library. Did you actually read the code?
Anyway. I changed the default in the library. And these values are the default values in the library.

Comment (ben-jones) [inline centinel/primitives/dnslib.py]: What I am asking is, do you have a response to my argument for why the timeout was initially set so high? If you agree to leaving the timeout at its original default, then we can cut your changes to the default value from this library. If you don't agree with the original timeout value, then I would ask you to respond to the previous argument.

Comment (rpanah) [inline centinel/primitives/dnslib.py]: > I think that 2 seconds should be sufficient, but is not guaranteed to be. Some regions (think Africa) typically have RTTs in the range of 400-500ms. If this is a query over TCP (common for EDNS options), then you have two RTTs for the query, which could easily be over 1 second. It would be odd, but I could totally see congestion issues leading to longer timeouts. In our case, the timeout is defined for the socket, but I think it is possible for the timeout to exceed 2 seconds.

I agree that there are some countries that have RTTs that high, but I don't agree that we should make it so much longer for every country everywhere else. We could do country-specific experiments for cases where RTTs happens to be that high, but for others, I think 2 seconds is pretty reasonable.

> Ultimately, the average or typical case really doesn't matter in this case. If we are going to make that claim that a DNS request is being dropped based on this timeout, we cannot settle for an inconclusive measurement-> 2 second is not conclusive enough for me. In these odd cases, we need to wait longer to differentiate packet drops and longer delays. If performance is a concern, then I also built a parallel version of this, so it really doesn't matter.

2 seconds is not inconclusive in **most** cases. Other cases with higher RTTs can be studied separately. As for concurrency, we don't want to have multiple DNS queries issued at the same time because we are recording pcaps for each URL individually. Doing everything concurrently would make pcaps unclean.

Comment (rpanah) [inline centinel/primitives/dnslib.py]: > do you have a response to my argument

I answered from the bottom to the top. :-)

File: centinel/experiments/baseline.py (hunk 6 above)
Comment (rpanah) [inline centinel/experiments/baseline.py]: You're right. I'll switch to `csv` library.

File: centinel/experiments/baseline.py (hunk 7 above)
Comment (rpanah) [inline centinel/experiments/baseline.py]: I'll refactor.

File: centinel/experiments/baseline.py (hunk 4 above)
Comment (rpanah) [inline centinel/experiments/baseline.py]: There's only one `if` statement. It `continue`s at the end.

File: centinel/primitives/dnslib.py (hunk 10 above)
Comment (ben-jones) [inline centinel/primitives/dnslib.py]: > I agree that there are some countries that have RTTs that high, but I don't agree that we should make it so much longer for every country everywhere else. We could do country-specific experiments for cases where RTTs happens to be that high, but for others, I think 2 seconds is pretty reasonable.

I buy that we might want to do country specific RTTs, but only if we didn't have the ability to do concurrency (more on that in a second). Otherwise, setting the largest possible value isn't that big a deal.

> 2 seconds is not inconclusive in most cases. Other cases with higher RTTs can be studied separately. As for concurrency, we don't want to have multiple DNS queries issued at the same time because we are recording pcaps for each URL individually. Doing everything concurrently would make pcaps unclean.

Concurrency shouldn't be an issue for us either at the collection or the analysis stage. If we assume that we are on RPIs, which we previously have, then we can safely collect all traffic to target ports without an risk of collecting user data. For example, we could collect all the data or just collect all of the packets destined for port 53 to get all DNS traffic. Additionally, concurrency won't be an issue at the analysis stage because a) bro is awesome for analyzing pcaps will give us a log of all of the DNS requests and responses in the pcap and b) it is really easy to correlate DNS requests and responses. DNS packets contain a query ID field that *must* match between the query and response and the responses often contain the original question, making identification even easier.

Basically, I don't see a strong argument for not using concurrency. If we have concurrency, then there is no reason to reduce the timeout.

File: centinel/experiments/baseline.py (hunk 8 above)
Comment (rpanah) [inline centinel/experiments/baseline.py]: You're right. It might break. I'll fix it.

File: centinel/primitives/dnslib.py (hunk 10 above)
Comment (rpanah) [inline centinel/primitives/dnslib.py]: Though I understand that we could still extract useful information from a pcap, I'd still argue that having one clean pcap for each URL makes the results easier to use later. A giant pcap that contains data for ~1000 URLs is going to be unwieldy and unnecessary. Uploading one big pcap file is also more challenging in poor network conditions than a uploading one small pcap at a time.

Comment (ben-jones) [inline centinel/primitives/dnslib.py]: We already weren't going to have one clean pcap-> the capture starts before we run the experiment. Also, a single pcap file will be better for storage-> if you store a lot of data, then larger files are better. I don't think we are talking about a large enough pcap for upload problems to be significant either.

Comment (rpanah) [inline centinel/primitives/dnslib.py]: In addition to that, 10 seconds is too long in any case. I think 5 seconds is more reasonable if we were to increase the timeout.

Comment (rpanah) [inline centinel/primitives/dnslib.py]: > We already weren't going to have one clean pcap

We are, to some extent. The main usecase for this is to run on dedicated hardware periodically.

> if you store a lot of data, then larger files are better

Not true. We are doing a `sync` once every hour. Smaller files are transferred more reliably that way.

> I don't think we are talking about a large enough pcap for upload problems to be significant either.

Also not true as I ran into this issue while running the benchmark on test nodes in the field.

**Comment by ben-jones** [comment]
@rpanah, I have responded to each of your points inline, but I'm going to leave an overall summary here. Basically, we gain a lot of value (speed, robust measurements) from leaving the timeout high and doing concurrent requests. I don't find your arguments compelling enough to reduce that value, though I welcome further critiques.

> We are, to some extent. The main usecase for this is to run on dedicated hardware periodically.

I disagree with this. What is a clean pcap to you? By default, we collect everything, then we can strip off what we don't need. This means that any updates, command and control traffic, and anything else are going to be caught in the pcap, which won't be clean (meaning that other concurrent traffic will be captured by centinel.) We can add code to get rid of this, but it's probably not worth it.

Also, I don't buy that this is necessary. What is the value in having a "clean" pcap? Why would that make analysis any easier? To do a robust analysis, we are going to have to apply higher level information (e.g. matching DNS queries and responses by query ID), so it shouldn't make a difference how much stuff is being requested concurrently.

> Not true. We are doing a sync once every hour. Smaller files are transferred more reliably that way.
> Also not true as I ran into this issue while running the benchmark on test nodes in the field.

To clarify, right now we are doing a single big pcap that is collected for the entire experiment (all primitives contained within the pcap). This capture file is going to be the same size regardless of whether we make the request concurrently. Furthermore, if size is an issue, we can split the pcap up using [splitcap](http://wiki.wireshark.org/Tools) or tcpdump.

Regardless, my point was that having a large file is better for the filesystem and this is going to be a problem for us on the backend. We shouldn't be talking about that much data though, so I don't buy that this is a problem that we need to solve. What are the sizes of the pcaps you were having problems with, what were the problems, and why do you think the problems are related to the size of the file?

File: centinel/primitives/dnslib.py (hunk 10 above)
Comment (ben-jones) [inline centinel/primitives/dnslib.py]: @rpanah, I seem to be having some issues with github's commenting system, so in case you didn't see my comment in the main thread, I am duplicating it here. Let me know what you think and please respond with data if available.

@rpanah, I have responded to each of your points inline, but I'm going to leave an overall summary here. Basically, we gain a lot of value (speed, robust measurements) from leaving the timeout high and doing concurrent requests. I don't find your arguments compelling enough to reduce that value, though I welcome further critiques.

>We are, to some extent. The main usecase for this is to run on dedicated hardware periodically.

I disagree with this. What is a clean pcap to you? By default, we collect everything, then we can strip off what we don't need. This means that any updates, command and control traffic, and anything else are going to be caught in the pcap, which won't be clean (meaning that other concurrent traffic will be captured by centinel.) We can add code to get rid of this, but it's probably not worth it.

Also, I don't buy that this is necessary. What is the value in having a "clean" pcap? Why would that make analysis any easier? To do a robust analysis, we are going to have to apply higher level information (e.g. matching DNS queries and responses by query ID), so it shouldn't make a difference how much stuff is being requested concurrently.

>Not true. We are doing a sync once every hour. Smaller files are transferred more reliably that way.
    Also not true as I ran into this issue while running the benchmark on test nodes in the field.

To clarify, right now we are doing a single big pcap that is collected for the entire experiment (all primitives contained within the pcap). This capture file is going to be the same size regardless of whether we make the request concurrently. Furthermore, if size is an issue, we can split the pcap up using splitcap or tcpdump.

Regardless, my point was that having a large file is better for the filesystem and this is going to be a problem for us on the backend. We shouldn't be talking about that much data though, so I don't buy that this is a problem that we need to solve. What are the sizes of the pcaps you were having problems with, what were the problems, and why do you think the problems are related to the size of the file?

File: centinel/experiments/baseline.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                    tls_results[domain_name] = tls_result
+                except Exception as exp:
+                    logging.info("%s: TLS certfiticate download failed: %s" %
+                                 (domain_name, exp))
+                    tls_results[domain_name] = { "exception" : str(exp) }
+
+            # DNS Lookup
+            logging.info("%s: DNS" % (domain_name))
+            try:
+                dns_results[domain_name] = dnslib.lookup_domain(domain_name)
```
Comment (ben-jones) [inline centinel/experiments/baseline.py]: @rpanah, what are we doing about concurrency here? Are we going to make multiple DNS requests in parallel with the lookup_domains call? Are you going to create a concurrent HTTP primitive?

File: centinel/primitives/dnslib.py (hunk 10 above)
Comment (ben-jones) [inline centinel/primitives/dnslib.py]: To conclude this discussion, can you change the timeout value back up to a larger value and we will figure out the concurrency stuff?

**Comment by ben-jones** [comment]
@rpanah, I'm generally ok with the code, but we still haven't resolved what we are doing with concurrency. Are you going to add concurrency to the HTTP primitive and do concurrent HTTP and DNS lookups? What is your timeline for updating that?

File: centinel/primitives/dnslib.py (hunk 10 above)
Comment (feamster) [inline centinel/primitives/dnslib.py]: +1

On March 6, 2015 4:49:53 PM EST, Ben Jones <notifications@github.com> wrote:
>> @@ -15,7 +15,7 @@ def get_ips(host, nameserver=None, record="A"):
>>      return lookup_domain(host, nameservers=nameservers,
>rtype=record)
>>  
>>  
>> -def lookup_domain(domain, nameservers=[], rtype="A", timeout=10):
>> +def lookup_domain(domain, nameservers=[], rtype="A", timeout=2):
>
>To conclude this discussion, can you change the timeout value back up
>to a larger value and we will figure out the concurrency stuff?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/iclab/centinel/pull/124/files#r25981176


-Nick

File: centinel/experiments/baseline.py (hunk 11 above)
Comment (rpanah) [inline centinel/experiments/baseline.py]: That's what we concluded from our discussion this week. I'll make both of these concurrent.

Comment (ben-jones) [inline centinel/experiments/baseline.py]: Awesome. I assume then that you are going to write and push the new primitives, then update the baseline experiments? I am ok with merging the rest of the baseline stuff now, then refactoring with another pull request as long as we change the timeout value.

Comment (rpanah) [inline centinel/experiments/baseline.py]: > I am ok with merging the rest of the baseline stuff now, then refactoring with another pull request as long as we change the timeout value.

Alright. I'll restore the timeout value and push the rest in another pull request.

Comment (rpanah) [inline centinel/experiments/baseline.py]: @ben-jones, 'Tis done. Thou shalt merge.

**Comment by ben-jones** [comment]
@rpanah, per our discussion, I am going to merge this without the concurrent code, but we will refactor after we make a concurrent HTTP primitive.
